hilti/src/rt/tests/library.cc

```diff
@@ -0,0 +1,225 @@
+    std::streambuf* _old = nullptr;
+    std::ostream* _stream = nullptr;
+};
+
+TEST_SUITE_BEGIN("Library");
+
+TEST_CASE("construct") {
+    Library _(dummy); // Does not throw
+    CHECK_THROWS_WITH_AS(Library("/does/not/exist"), "no such library: \"/does/not/exist\"", const EnvironmentError&);
+
+    SUBCASE("TMPDIR does not exist") {
+        Env _("TMPDIR", "/does/not/exist");
+        try {
+            Library _(dummy);
+            FAIL("expected exception not thrown");
+        } catch ( const EnvironmentError& e ) {
+            CAPTURE(e.description());
+            CHECK(startsWith(e.description(),
+                             fmt("could not add library \"%s\": could not create temporary file: ", dummy.c_str())));
+        } catch ( ... ) {
+            FAIL("unexpected exception thrown");
+        }
+    }
+
+    SUBCASE("cannot write TMPDIR") {
+        // NOTE: This test likely does not work if run as `root` as `root`
+        // can probably create files even in read-only directories.
+        TemporaryDirectory tmpdir;
+        std::filesystem::permissions(tmpdir.path(), std::filesystem::perms::none);
+        Env _("TMPDIR", tmpdir.path().c_str());
+
+        try {
+            Library _(dummy);
+            FAIL("expected exception not thrown");
+        } catch ( const EnvironmentError& e ) {
+            CAPTURE(e.description());
+            CHECK(startsWith(e.description(),
+                             fmt("could not add library \"%s\": could not create temporary file in ", dummy.c_str())));
+        } catch ( ... ) {
+            FAIL("unexpected exception thrown");
+        }
+    }
+}
+
+TEST_CASE("destruct") {
+    TemporaryDirectory tmp;
+    Env _("TMPDIR", tmp.path().c_str());
+
+    auto library = std::unique_ptr<Library>(new Library(dummy));
+
+    // Recursively change the permissions of the directory
+    // so the temporary library cannot be removed anymore.
+    //
+    // NOTE: This test likely does not work if run as `root` as `root`
+    // can probably create files even in read-only directories.
+    for ( const auto& entry : std::filesystem::recursive_directory_iterator(tmp.path()) )
+        std::filesystem::permissions(entry, std::filesystem::perms::none);
+    std::filesystem::permissions(tmp.path(), std::filesystem::perms::none);
+
+    CaptureIO cerr(std::cerr);
+
+    CHECK_NOTHROW(library.reset());
+    CHECK_NE(cerr.str().find("could not remove library"), std::string::npos);
+}
+
+TEST_CASE("open") {
+    SUBCASE("success") { // Symbol not loaded, yet.
+        REQUIRE_FALSE(symbol("foo").hasValue());
+
+        {
+            // Creating `Library` instance does not load it automatically.
+            Library library(dummy);
+            CHECK_FALSE(symbol("foo").hasValue());
+
+            // Explicitly opening the `Library` loads it so the symbol can be found.
+            CHECK_EQ(library.open(), Nothing());
+            CHECK(symbol("foo").hasValue());
```

Comment:
one more test that would be useful if easy to add: if we have two libraries that provide the same symbol, we want to (1) be able to open them both, and (2) still get access to each's version of the symbol. We need this for the new HLTO versioning: same symbol, but library-specific content.

Reply:
This is currently not possible since `Library` does not hold handles to libraries it wraps and users instead call `::dlsym` which might return a unexpected symbol if multiple symbols with the same name exist.

I'll add a member function to obtain a symbol from a specific `Library`; we'd still need to update calling code to use that instead of global `::dlsym` which would also require them to hold on to instances of `Library` (once we fix that we should also add `::dlclose` to `Library` destructors).